lib/livebook_web/plugs/proxy_plug.ex

```diff
@@ -0,0 +1,60 @@
+defmodule LivebookWeb.ProxyPlug do
+  @behaviour Plug
+  import Plug.Conn
+
+  alias Livebook.{App, Apps, Session, Sessions}
+  alias LivebookWeb.NotFoundError
+
+  @impl true
+  def init(opts), do: opts
+
+  @impl true
+  def call(%{path_info: ["sessions", id, "proxy" | path_info]} = conn, _opts) do
+    session = fetch_session!(id)
+    pid = fetch_kino_proxy!(session)
+    conn = prepare_conn(conn, path_info, ["sessions", id, "proxy"])
+    {conn, _} = Kino.Proxy.serve(pid, conn)
+
+    halt(conn)
+  end
+
+  def call(%{path_info: ["apps", slug, "proxy" | path_info]} = conn, _opts) do
+    app = fetch_app!(slug)
+    id = App.get_session_id(app.pid)
```

Comment:
@josevalim should we validate that the app is single-session? Also, `auto_shutdown_ms` would be an issue for the first request.

Reply:
Good call. We should also support multi-session apps. What would be the URL in this case?

    def call(%{path_info: ["apps", _slug, session_id, "proxy" | path_info]} = conn, _opts) do

?

Reply:
Technically each app session has that URL, for single-session we redirect the user to that specific URL right away.

The main concern is hitting the app at start, while the proxy is not started yet. So your idea is that the user would start the multi-session app session, perhaps enter some data, and then they could query that session's proxy?

Reply:
Yes, exactly. We need to document on `Kino.Proxy.listen` that:

1. For single session apps, you can access it at `/apps/slug/proxy` for the other apps and session, you need the session URL
2. You must not enable automatic shutdowns, we don't automatically start apps on demand at the moment

Reply:
> we don't automatically start apps on demand at the moment

With the current code we do, but they need time to setup :D So should we change it to only use existing single-session and error otherwise?

Reply:
yes, i'd say so

Reply:
Or we add the functionality to block until they boot, if it is not a lot of work. Which one do you prefer?

Reply:
Until boot, as in, we wait for it to become fully executed, and in case of error/interrupt we respond with an error? Shouldn't bee too much work, I can have a look once this is merged.

If we do that, then we also want to pass the user when creating multi-session sessions.

Reply:
Gah, that's annoying for multi-session apps. So we either require it to be running OR we don't support booting behind the scenes for multi sessions. Thoughts?

Reply:
Well, for multi session you always need to pass the session ID, which means we can't start it, so that's problem solved. So overall:

  * For single session apps, you can access it at /apps/slug/proxy, it will be automatically started
  * For multi session apps, you MUST access at /apps/slug/session_id/proxy, it is not automatically started

Reply:
> Well, for multi session you always need to pass the session ID, which means we can't start it

I mean, we could create one on every hit to `/apps/slug/proxy` (in which case you probably want `auto_shutdown_ms` 5s), but it's probably not that useful. Multi is usually for interrupts anyway.

So in the end you would require the app to be fully executed, and we support automatic starts on `/apps/slug/proxy` for single-session apps?

Reply:
> So in the end you would require the app to be fully executed, and we support automatic starts on /apps/slug/proxy for single-session apps?

100% agreed.

So what about this: for this PR, let's only support "/apps/:slug/:session_id/proxy" and then we add support for "/apps/:slug/proxy" with automatic boot in a later PR?

Reply:
No strong opinion on what to include, just note that even with direct URL we still need to await the app session to be executed.

Reply:
@jonatanklosko why do we need to wait? shouldn't we just assume that it is already running?

Reply:
I mean, if we do the work to wait on the automatic boot, we should as well wait in the other cases too.

Reply:
Hrm... I am worried someone may have a never completing app and then those would never work under any URL and having a distinction between them could be helpful? I don't know.

Reply:
Like, given we need to have a session_id, it means you need to manually go and boot it up, I think it is fine to assume it is running.

Reply:
Anyway, we can decide this later. Let's keep this on the path to merging now. I will resolve the conversation and drop a more precise feedback for Ale.

Reply:
> it means you need to manually go and boot it up

For single session it's just started on deployment, so there's a race condition when someone hits the app too soon.

> I am worried someone may have a never completing app

I would error on errors and interrupts, if it takes forever to execute, then it's actually something the user should know and fix, no?

Reply:
I guess it is fair and the request will eventually timeout anyway. But yeah, let's implement this later. :)